### DISCUSSION TITLE
Permission denied in git status command

### DISCUSSION THREAD

**Discussion by hyeom-wisc** [idaholab/moose#16338] [Q&A Getting Started] [answered]
Hello,
I am a new user in Moose code and moving forward step by step by asking questions in this forum. Thank you so much all! 

I am working on tutorial01. I was able to run a new input file and see the output through peacock. It took a week for me to get this stage. However, unfortunately, I have some issue to run "git status" It shows: 

```
(moose) hyeom@ERB914-7920:~/projects/babbler$ git status
fatal: .git/index: index file open failed: Permission denied
```

In GitHub desktop, it says "last fetched just now on" but I cannot Pull in Git tab in Atom. I think I messed it up somehow. Can you please provide some clues to resolve this? Thanks. Sung

**Comment by milljm** [comment, marked as answer]
Permission denied errors usually are the result of running commands with `sudo`.
Can you perform a long listing while 'babbler'?

```bash
ls -latrh ~/projects/babbler
```

Thanks!

**Reply by hyeom-wisc** [reply]
Dear Jason,
I typed the command and the result is shown below. Can you see some issue? Thanks. Sung.

```
(moose) hyeom@ERB914-7920:~/projects/babbler$ ls -latrh ~/projects/babbler
total 476K
-rw-r--r-- 1 hyeom hyeom  26K Nov 24 12:17 LICENSE
drwxr-xr-x 1 hyeom hyeom 4.0K Nov 24 12:17 include
drwxr-xr-x 1 hyeom hyeom 4.0K Nov 24 12:17 scripts
-rw-r--r-- 1 hyeom hyeom  183 Nov 24 12:17 README.md
-rw-r--r-- 1 hyeom hyeom 2.1K Nov 24 12:17 Makefile
-rwxr-xr-x 1 hyeom hyeom  470 Nov 24 12:17 run_tests
-rw-r--r-- 1 hyeom hyeom   86 Nov 24 12:17 testroot
drwxr-xr-x 1 hyeom hyeom 4.0K Nov 24 12:17 unit
-rw-r--r-- 1 hyeom hyeom  738 Nov 24 12:17 .clang-format
-rw-r--r-- 1 hyeom hyeom 3.2K Nov 24 12:17 .gitignore
drwxr-xr-x 1 hyeom hyeom 4.0K Nov 24 12:25 build
drwxr-xr-x 1 hyeom hyeom 4.0K Nov 24 12:26 test
drwxr-xr-x 1 hyeom hyeom 4.0K Nov 24 12:26 .libs
drwxr-xr-x 1 hyeom hyeom 4.0K Nov 27 11:51 ..
drwxr-xr-x 1 hyeom hyeom 4.0K Nov 28 15:06 src
drwxr-xr-x 1 hyeom hyeom 4.0K Nov 28 15:07 lib
drwxr-xr-x 1 hyeom hyeom 4.0K Nov 28 15:08 .
-rwxr-xr-x 1 hyeom hyeom 426K Nov 28 15:08 babbler-opt
-rw-r--r-- 1 hyeom hyeom 7.7K Nov 28 15:09 .previous_test_results.json
drwxr-xr-x 1 hyeom hyeom 4.0K Nov 30 09:53 doc
drwxr-xr-x 1 hyeom hyeom 4.0K Nov 30 09:53 .git
drwxr-xr-x 1 hyeom hyeom 4.0K Nov 30 11:18 problems
```

**Reply by dschwen** [reply]
Better do a `ls -latrh ~/projects/babbler/.git`

**Reply by dschwen** [reply]
As Jason stated, you've probably run `git` with `sudo` at some point, changing the ownership of some internal git files to `root`. You can fix that by doing

```
sudo chown -R hyeom:hyeom ~/projects/babbler
```

and then never use `sudo` again when working with MOOSE :-)

**Reply by milljm** [reply]
maybe there are aspects of git which are not owned by you... can you perform the following:

```bash
ls -latrh ~/projects/babler/.git/
```

Thanks!

EDIT: woops. Didn't see that Daniel already replied.

**Reply by hyeom-wisc** [reply]
Dear Daniel,
It showed the below after typing ls -latrch ~/projects/babber/.git

(moose) hyeom@ERB914-7920:~/projects/babbler$ ls -latrh ~/projects/babbler/.git
total 4.0K
drwxr-xr-x 1 hyeom hyeom 4.0K Nov 24 12:17 branches
-rw-r--r-- 1 hyeom hyeom   73 Nov 24 12:17 description
drwxr-xr-x 1 hyeom hyeom 4.0K Nov 24 12:17 hooks
drwxr-xr-x 1 hyeom hyeom 4.0K Nov 24 12:17 info
-rw-r--r-- 1 hyeom hyeom   23 Nov 24 12:17 HEAD
drwxr-xr-x 1 hyeom hyeom 4.0K Nov 24 12:23 logs
drwxr-xr-x 1 hyeom hyeom 4.0K Nov 24 12:25 refs
---------- 1 hyeom hyeom   41 Nov 28 14:32 ORIG_HEAD
drwxr-xr-x 1 hyeom hyeom 4.0K Nov 28 14:37 objects
---------- 1 hyeom hyeom 2.7K Nov 28 14:37 index
---------- 1 hyeom hyeom   15 Nov 28 14:37 COMMIT_EDITMSG
-rw-r--r-- 1 hyeom hyeom  262 Nov 28 15:02 config
drwxr-xr-x 1 hyeom hyeom 4.0K Nov 28 15:08 ..
drwxr-xr-x 1 hyeom hyeom 4.0K Nov 30 09:53 .
---------- 1 hyeom hyeom    0 Nov 30 11:29 FETCH_HEAD

After then, I tried the sudo chown and then typed git status. However, it still says "fatal: .git/index: index file open failed: Permission denied" 

Sung

**Reply by hyeom-wisc** [reply]
Daniel and Jason,

As you said, I think some of the files (please see above) are not owned by me. Can you please let me know how to change the ownership? 

Sincerely,
Sung

**Reply by GiudGiud** [reply]
Hi Sung

Type: chmod +rw file_name (or * to get them all at once), in the .git directory, to change file read/write priviledges.

Best,
Guillaume

**Reply by hyeom-wisc** [reply]
Guillaume,
Thanks. Now I can run "git status" after changing the ownership of .git folder. After then I typed "git add problems" and it showed another error. There are many permission issues... Sung

(moose) hyeom@ERB914-7920:~/projects/babbler$ git add problems
error: insufficient permission for adding an object to repository database .git/objects
error: problems/pressure_diffusion.i: failed to insert into database
error: unable to index file 'problems/pressure_diffusion.i'
fatal: updating files failed

**Reply by GiudGiud** [reply]
Maybe try changing the permissions recursively, for all files in the folder at once. 
So 'chmod +rw --recursive ~/projects/babbler'

**Reply by hyeom-wisc** [reply]
Guillaume,
Thanks. I run it but it showed many "cannot access" error messages.

(moose) hyeom@ERB914-7920:~/projects/babbler$ chmod +rw --recursive ~/projects/babbler
chmod: cannot access '/home/hyeom/projects/babbler/.git/objects/01/6580497728634160a6541a6342e2ba07000696': Permission denied
chmod: cannot access '/home/hyeom/projects/babbler/.git/objects/02/ae105e1a8d4370e00d1ce6bba5a072e299ad19': Permission denied
chmod: cannot access '/home/hyeom/projects/babbler/.git/objects/03/f7cb153298d959ee5bbc4493ad80c87341a684': Permission denied
chmod: cannot access '/home/hyeom/projects/babbler/.git/objects/0b/3f587194e53ab2c855ea66dac23662289fec90': Permission denied
chmod: cannot access '/home/hyeom/projects/babbler/.git/objects/17/b27300a29aad760be364281469b72f0ff4d1d7': Permission denied
chmod: cannot access '/home/hyeom/projects/babbler/.git/objects/34/154808f950bd18324a8b5ed914fc737c4dd931': Permission denied
chmod: cannot access '/home/hyeom/projects/babbler/.git/objects/34/58c0efe9c0fe75be67df29bac8601f8a28c927': Permission denied
chmod: cannot access '/home/hyeom/projects/babbler/.git/objects/3a/13df19eef2cc5b3df81d920175b31c411ca88c': Permission denied
chmod: cannot access '/home/hyeom/projects/babbler/.git/objects/3d/ffbd86b56f30d464d543ac4542443a1fc0a04e': Permission denied
chmod: cannot access '/home/hyeom/projects/babbler/.git/objects/47/d8ee3d8a753c186f5aed84296b42661f9e3028': Permission denied
chmod: cannot access '/home/hyeom/projects/babbler/.git/objects/62/c62ed143f8a6868635ac82eef6a66791957e03': Permission denied
chmod: cannot access '/home/hyeom/projects/babbler/.git/objects/79/201b6137aa03fe34b8c5867044a2f187fb6eea': Permission denied
chmod: cannot access '/home/hyeom/projects/babbler/.git/objects/87/249798ab181373d276a068dd3aec85d6d8ac4d': Permission denied
chmod: cannot access '/home/hyeom/projects/babbler/.git/objects/8c/aaf328a9718c6057e7f587a824f05af242d2e2': Permission denied
chmod: cannot access '/home/hyeom/projects/babbler/.git/objects/96/027c6336b3aecb7b9eea2a09a81f5df08e3efb': Permission denied
chmod: cannot access '/home/hyeom/projects/babbler/.git/objects/a3/2d71723da2ea0f4273943acd5192fbe1a717c0': Permission denied
chmod: cannot access '/home/hyeom/projects/babbler/.git/objects/b5/52b8e3d4fc5f233a5c9e8bf0d8e9bb400ed992': Permission denied
chmod: cannot access '/home/hyeom/projects/babbler/.git/objects/b5/6d12faeae50810d231b12b1f6b7b37cf9b0d0e': Permission denied
chmod: cannot access '/home/hyeom/projects/babbler/.git/objects/e1/8a5a2ec524dee700af30333b153710d391f471': Permission denied

**Reply by GiudGiud** [reply]
I'm not sure you need permission on those files.
Can you git add problems now?

If not, you may want to sudo chmod, but double check that the permissions are actually given to you

**Reply by GiudGiud** [reply]
> As Jason stated, you've probably run `git` with `sudo` at some point, changing the ownership of some internal git files to `root`. You can fix that by doing
> 
> ```
> sudo chown -R hyeom:hyeom ~/projects/babbler
> ```
> 
> and then never use `sudo` again when working with MOOSE :-)

Actually this should work

**Comment by hyeom-wisc** [comment]
Guillaume,
I am frustrated... I tried two things again.
1) 
(moose) hyeom@ERB914-7920:~/projects/babbler$ sudo chown -R hyeom:hyeom ~/projects/babbler
[sudo] password for hyeom:
2) 
(moose) hyeom@ERB914-7920:~/projects/babbler$ sudo chmod +rw --recursive ~/projects/babbler
3)
(moose) hyeom@ERB914-7920:~/projects/babbler$ git add problems
error: insufficient permission for adding an object to repository database .git/objects
error: problems/pressure_diffusion.i: failed to insert into database
error: unable to index file 'problems/pressure_diffusion.i'
fatal: updating files failed

It still showed the error message. It is not easy.

**Reply by hyeom-wisc** [reply]
Guillaume,

Interestingly, now the git status doesn't work again. I am sorry for taking your time on this issue.

(moose) hyeom@ERB914-7920:~/projects/babbler$ git status
fatal: failed to read object b56d12faeae50810d231b12b1f6b7b37cf9b0d0e: Permission denied

**Reply by GiudGiud** [reply]
The two commands you ran (chown -R ... and chmod --recursive ...) do similar things. -R is a short hand for --recursive.
The first one gives you ownership of the files, and the second read/write access

Can you check the permissions again? With ls -latrh. Maybe run that in ~/projects/babbler and ~/projects/babbler/.git

**Reply by GiudGiud** [reply]
If fixing the permissions is becoming too time consuming, I recommend you re-create the application, and only move the files that you modified to the new application. 
If this is your first commit, it should not take too long.

**Reply by hyeom-wisc** [reply]
Guillaume,
Please see the two ls -latrh commands. Can you see something strange?
1) 
(moose) hyeom@ERB914-7920:~/projects/babbler$ ls -latrh ~/projects/babbler
total 476K
-rw-r--r-- 1 hyeom hyeom  26K Nov 24 12:17 LICENSE
drwxr-xr-x 1 hyeom hyeom 4.0K Nov 24 12:17 include
drwxr-xr-x 1 hyeom hyeom 4.0K Nov 24 12:17 scripts
-rw-r--r-- 1 hyeom hyeom  183 Nov 24 12:17 README.md
-rw-r--r-- 1 hyeom hyeom 2.1K Nov 24 12:17 Makefile
-rwxr-xr-x 1 hyeom hyeom  470 Nov 24 12:17 run_tests
-rw-r--r-- 1 hyeom hyeom   86 Nov 24 12:17 testroot
drwxr-xr-x 1 hyeom hyeom 4.0K Nov 24 12:17 unit
-rw-r--r-- 1 hyeom hyeom  738 Nov 24 12:17 .clang-format
-rw-r--r-- 1 hyeom hyeom 3.2K Nov 24 12:17 .gitignore
drwxr-xr-x 1 hyeom hyeom 4.0K Nov 24 12:25 build
drwxr-xr-x 1 hyeom hyeom 4.0K Nov 24 12:26 test
drwxr-xr-x 1 hyeom hyeom 4.0K Nov 24 12:26 .libs
drwxr-xr-x 1 hyeom hyeom 4.0K Nov 27 11:51 ..
drwxr-xr-x 1 hyeom hyeom 4.0K Nov 28 15:06 src
drwxr-xr-x 1 hyeom hyeom 4.0K Nov 28 15:07 lib
drwxr-xr-x 1 hyeom hyeom 4.0K Nov 28 15:08 .
-rwxr-xr-x 1 hyeom hyeom 426K Nov 28 15:08 babbler-opt
-rw-r--r-- 1 hyeom hyeom 7.7K Nov 28 15:09 .previous_test_results.json
drwxr-xr-x 1 hyeom hyeom 4.0K Nov 30 09:53 doc
drwxr-xr-x 1 hyeom hyeom 4.0K Nov 30 11:18 problems
drwxr-xr-x 1 hyeom hyeom 4.0K Dec  1 11:29 .git
2)
(moose) hyeom@ERB914-7920:~/projects/babbler$ ls -latrh ~/projects/babbler/.git
total 8.0K
drwxr-xr-x 1 hyeom hyeom 4.0K Nov 24 12:17 branches
-rw-r--r-- 1 hyeom hyeom   73 Nov 24 12:17 description
drwxr-xr-x 1 hyeom hyeom 4.0K Nov 24 12:17 hooks
drwxr-xr-x 1 hyeom hyeom 4.0K Nov 24 12:17 info
-rw-r--r-- 1 hyeom hyeom   23 Nov 24 12:17 HEAD
drwxr-xr-x 1 hyeom hyeom 4.0K Nov 24 12:23 logs
drwxr-xr-x 1 hyeom hyeom 4.0K Nov 24 12:25 refs
-rw-r--r-- 1 hyeom hyeom   41 Nov 28 14:32 ORIG_HEAD
drwxr-xr-x 1 hyeom hyeom 4.0K Nov 28 14:37 objects
-rw-r--r-- 1 hyeom hyeom  262 Nov 28 15:02 config
drwxr-xr-x 1 hyeom hyeom 4.0K Nov 28 15:08 ..
-rw-r--r-- 1 hyeom hyeom    0 Nov 30 11:29 FETCH_HEAD
-rw-r--r-- 1 hyeom hyeom 2.7K Dec  1 10:26 index
-rw-r--r-- 1 hyeom hyeom 1.4K Dec  1 10:39 COMMIT_EDITMSG
drwxr-xr-x 1 hyeom hyeom 4.0K Dec  1 11:29 .

**Reply by hyeom-wisc** [reply]
I agree. I may remove the whole babbler directory (should I do that in github website?) and restart the tutorial again.

Sincerely,
Sung

**Reply by GiudGiud** [reply]
The permissions look fine. You are the owner of the files (thanks to chown) and you have read/write access on all these files (first set of rw- is for the owner). 

You can worry about removing the directory later (from both your github account and your computer). For now, maybe try restarting the tutorial without using sudo.

**Reply by hyeom-wisc** [reply]
Guillaume,
Thanks. I will restart the tutorial with a new directory (e.g., babbler2) and see what happens. 

Sincerely,
Sung
